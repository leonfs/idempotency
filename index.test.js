const {check, configure} = require('.');
const {put} = require('./lib/ddb');
const {exists, set} = require('./lib/cache');

jest.mock('./lib/ddb');
jest.mock('./lib/cache');

describe('idempotency', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    global.Date.now = jest.fn(() => 1000);
  });

  it('should return replay true if cached', async () => {
    exists.mockReturnValue(true);
    const {replay} = await check({id: 1});
    expect(replay).toEqual(true);

    expect(exists.mock.calls).toMatchSnapshot();
    expect(set.mock.calls).toMatchSnapshot();
    expect(put.mock.calls).toMatchSnapshot();
  });

  it('should return replay true if not cached, but in ddb', async () => {
    exists.mockReturnValue(false);
    put.mockRejectedValue({code: 'ConditionalCheckFailedException'});

    const {replay} = await check({id: 1});
    expect(replay).toEqual(true);

    expect(exists.mock.calls).toMatchSnapshot();
    expect(set.mock.calls).toMatchSnapshot();
    expect(put.mock.calls).toMatchSnapshot();
  });

  it('should return replay false if not cached, or not in ddb', async () => {
    exists.mockReturnValue(false);
    put.mockResolvedValue();

    const {replay} = await check({id: 1});
    expect(replay).toEqual(false);

    expect(exists.mock.calls).toMatchSnapshot();
    expect(set.mock.calls).toMatchSnapshot();
    expect(put.mock.calls).toMatchSnapshot();
  });

  it('should allow things to be configured', async () => {
    config = {
      tablename: 'TABLE',
      idField: 'ID',
      ttlField: 'TTL',
      ttl: 42,
    };

    exists.mockReturnValue(false);
    put.mockResolvedValue();

    await configure(config);
    await check({id: 1});

    expect(put.mock.calls).toMatchSnapshot();
  });

  /*
  it.skip('can configure', async () => {
    await configure({tablename: 'XX'});
    // TODO assert DDB calls
  });

  it.skip('should return {replay} correctly', async () => {
    const first = await check({id: 1});
    expect(first.replay).toBe(false);
    const second = await check({id: 1});
    expect(second.replay).toBe(true);
  });

  it('should call ddb correctly', async () => {
    const {replay} = await check({id: 3});
    expect(replay).toEqual(false);
  });
  */
});
