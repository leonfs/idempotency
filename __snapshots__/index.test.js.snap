// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`idempotency should allow things to be configured 1`] = `
Array [
  Array [
    Object {
      "ConditionExpression": "attribute_not_exists(ID)",
      "Item": Object {
        "ID": "1",
        "TTL": 61,
      },
      "Key": Object {
        "ID": "1",
      },
      "TableName": "TABLE",
    },
  ],
]
`;

exports[`idempotency should return replay false if not cached, or not in ddb 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`idempotency should return replay false if not cached, or not in ddb 2`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`idempotency should return replay false if not cached, or not in ddb 3`] = `
Array [
  Array [
    Object {
      "ConditionExpression": "attribute_not_exists(id)",
      "Item": Object {
        "id": "1",
        "ttl": 61,
      },
      "Key": Object {
        "id": "1",
      },
      "TableName": "idempotency",
    },
  ],
]
`;

exports[`idempotency should return replay true if cached 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`idempotency should return replay true if cached 2`] = `Array []`;

exports[`idempotency should return replay true if cached 3`] = `Array []`;

exports[`idempotency should return replay true if not cached, but in ddb 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`idempotency should return replay true if not cached, but in ddb 2`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`idempotency should return replay true if not cached, but in ddb 3`] = `
Array [
  Array [
    Object {
      "ConditionExpression": "attribute_not_exists(id)",
      "Item": Object {
        "id": "1",
        "ttl": 61,
      },
      "Key": Object {
        "id": "1",
      },
      "TableName": "idempotency",
    },
  ],
]
`;
